src/v/cloud_storage/partition_manifest.cc

```diff
@@ -1209,114 +1246,169 @@ void partition_manifest::serialize(std::ostream& out) const {
+        w.Int64(static_cast<int16_t>(meta.sname_format));
+    }
+    w.EndObject();
+}
+
+void partition_manifest::serialize_removed_segment_meta(
+  const lw_segment_meta& meta, serialization_cursor_ptr cursor) const {
+    // Here we are serializing all fields stored in 'lw_segment_meta'.
+    // The remaining fields are also added but they values are not
+    // significant.
+    vassert(
+      meta.segment_term != model::term_id{},
+      "Term id is not initialized, base offset {}",
+      meta.base_offset);
+    auto& w = cursor->writer;
+    auto sn = generate_local_segment_name(meta.base_offset, meta.segment_term);
+    w.Key(sn());
+    w.StartObject();
+    w.Key("size_bytes");
+    w.Int64(static_cast<int64_t>(meta.size_bytes));
+    w.Key("committed_offset");
+    w.Int64(meta.committed_offset());
+    w.Key("base_offset");
+    w.Int64(meta.base_offset());
+    if (meta.ntp_revision != _rev) {
         vassert(
-          meta.segment_term != model::term_id{},
-          "Term id is not initialized, base offset {}",
-          meta.base_offset);
-        auto sn = generate_local_segment_name(
-          meta.base_offset, meta.segment_term);
-        w.Key(sn());
-        w.StartObject();
-        w.Key("size_bytes");
-        w.Int64(static_cast<int64_t>(meta.size_bytes));
-        w.Key("committed_offset");
-        w.Int64(meta.committed_offset());
-        w.Key("base_offset");
-        w.Int64(meta.base_offset());
-        if (meta.ntp_revision != _rev) {
-            vassert(
-              meta.ntp_revision != model::initial_revision_id(),
-              "ntp {}: missing ntp_revision for segment {} in the manifest",
-              _ntp,
-              sn);
-            w.Key("ntp_revision");
-            w.Int64(meta.ntp_revision());
-        }
-        if (meta.archiver_term != model::term_id::min()) {
-            w.Key("archiver_term");
-            w.Int64(meta.archiver_term());
-        }
-        w.Key("segment_term");
-        w.Int64(meta.segment_term());
-        w.EndObject();
-    };
-    if (!_segments.empty()) {
+          meta.ntp_revision != model::initial_revision_id(),
+          "ntp {}: missing ntp_revision for segment {} in the manifest",
+          _ntp,
+          sn);
+        w.Key("ntp_revision");
+        w.Int64(meta.ntp_revision());
+    }
+    if (meta.archiver_term != model::term_id::min()) {
+        w.Key("archiver_term");
+        w.Int64(meta.archiver_term());
+    }
+    w.Key("segment_term");
+    w.Int64(meta.segment_term());
+    w.EndObject();
+}
+
+void partition_manifest::serialize_segments(
+  partition_manifest::serialization_cursor_ptr cursor) const {
+    vassert(cursor->segments_done == false, "Segments are already serialized");
+    if (_segments.empty()) {
+        cursor->next_offset = model::offset{};
+        cursor->segments_done = true;
+        return;
+    }
+    auto& w = cursor->writer;
+    if (cursor->next_offset == model::offset{} && !_segments.empty()) {
```

Comment:
Is it possible `_segments` becomes emptied in between calls to `serialize_segments()`, causing us to serialize and write out a prefix of the manifest rather than a full or empty manifest? Curious about what concurrent operations are expected (would be nice to include somewhere as a comment, maybe around the cursor class?)